### PR TITLE
ROX-21198: Enrich deployments with Network Policies in Central

### PR DESCRIPTION
**Pull request by Maddosaurus**
## Description

This PR extends the capabilities Central has when `roxctl deployment check` is called.
Now Central is enriching a Deployment with Network Policies that would apply to it, were it deployed.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] ~~Evaluated and added CHANGELOG entry if required~~
- [ ] ~~Determined and documented upgrade steps~~
- [ ] ~~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed
1. Deploy to any cluster
1. Set up a new API token in Central and export it for roxctl to use
1. Set Central loglevel to debug
1. Use `roxctl` from PR https://github.com/stackrox/stackrox/pull/9110 to check a deployment (see addendum):
    - `./roxctl deployment check --cluster remote --namespace stackrox --file ./nginx.yaml`
1. Observe logline in Central with `Determined applied network policies for deployment nginx-deployment: ...` listing empty Network Policies
1. Apply the yaml to the Namespace (mainly creating the Network Policy): 
    - `kubectl -n stackrox apply -f ./nginx.yaml`
1. Observe logline in Central with `Determined applied network policies for deployment nginx-deployment: ...` **listing the Network Policy**

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

---

Addendum: nginx.yaml
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  namespace: stackrox
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80

---

apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: multi-port-egress
  namespace: stackrox
spec:
  podSelector:
    matchLabels:
      app: nginx
  policyTypes:
    - Ingress
  ingress:
    - from:
        - ipBlock:
            cidr: 10.0.0.0/24
      ports:
        - protocol: TCP
          port: 32000
          endPort: 32768
```

### CODE REVIEW

File: central/detection/service/service_impl.go [hunk 1]
```diff
@@ -220,10 +226,19 @@ func (s *serviceImpl) enrichAndDetect(ctx context.Context, enrichmentContext enr
 		EnforcementOnly: enrichmentContext.EnforcementOnly,
 	}
 
+	var appliedNetpols *augmentedobjs.NetworkPoliciesApplied
+	appliedNetpols, err = s.getAppliedNetpolsForDeployment(ctx, enrichmentContext, deployment)
+	if err != nil {
+		log.Warnf("Could not determine applied network policies for deployment %s. Continuing with deployment enrichment.", deployment.GetName())
```
Comment (vikin91) [inline central/detection/service/service_impl.go:232]: Would maybe `s/determine/find/` make this line a bit clearer?

File: central/detection/service/service_impl.go [hunk 2]
```diff
@@ -220,10 +226,19 @@ func (s *serviceImpl) enrichAndDetect(ctx context.Context, enrichmentContext enr
...
+	}
+
+	log.Debugf("Determined applied network policies for deployment %s: %+v", deployment.GetName(), appliedNetpols)
```
Comment (vikin91) [inline central/detection/service/service_impl.go:235]: This needs to be in an `else` branch, otherwise we will see both log lines: the one from l. 232 and this one together.

File: central/detection/service/service_impl.go [hunk 3]
```diff
@@ -239,6 +254,15 @@ func (s *serviceImpl) enrichAndDetect(ctx context.Context, enrichmentContext enr
 	}, nil
 }
 
+func (s *serviceImpl) getAppliedNetpolsForDeployment(ctx context.Context, enrichmentContext enricher.EnrichmentContext, deployment *storage.Deployment) (*augmentedobjs.NetworkPoliciesApplied, error) {
+	storedPolicies, err := s.netpols.GetNetworkPolicies(ctx, enrichmentContext.ClusterID, enrichmentContext.Namespace)
+	if err != nil {
+		return nil, errox.InvalidArgs.New("failed to look up network policies for deployment").CausedBy(err)
```
Comment (vikin91) [inline central/detection/service/service_impl.go:260]: look up, determine, find - would be cool to pick one and stick to it ;)

File: central/detection/service/service_impl.go [hunk 4]
```diff
@@ -352,6 +376,11 @@ func (s *serviceImpl) DetectDeployTimeFromYAML(ctx context.Context, req *apiV1.D
 		eCtx.ClusterID = clusterID
 	}
 
+	ns := req.GetNamespace()
+	if ns != "" {
+		eCtx.Namespace = ns
+	}
```
Comment (vikin91) [inline central/detection/service/service_impl.go:382]: Maybe a bit more compact, like that?

```suggestion
	if ns := req.GetNamespace(); ns != "" {
		eCtx.Namespace = ns
	}
```

File: central/detection/service/service_impl_test.go [hunk 5]
```diff
@@ -685,3 +690,23 @@ func TestFetchOptionFromRequest(t *testing.T) {
 		})
 	}
 }
+
+func TestGetAppliedNetpolsForDeployment(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	mockNetpolDS := networkPolicyMockStore.NewMockDataStore(mockCtrl)
+	mockNetpolDS.EXPECT().GetNetworkPolicies(gomock.Any(), gomock.Any(), gomock.Any()).Return(
```
Comment (vikin91) [inline central/detection/service/service_impl_test.go:697]: You know the cluster ID (2nd param) and namespace (3rd param), so maybe let's try to change `gomock.Any()` to something more concrete? 
In its current form, this mock will always return the contents of `Return(...)` block. If we specify those params, then the `Return(...)` block will be only returned when those parameters would match the call.
```suggestion
	clusterID := uuid.NewV4().String()
	mockNetpolDS.EXPECT().GetNetworkPolicies(gomock.Any(), clusterID, "ns").Return(
```

and then in 705 use the cluster id from the var `clusterID`.

File: central/detection/service/service_impl_test.go [hunk 6]
```diff
@@ -685,3 +690,23 @@ func TestFetchOptionFromRequest(t *testing.T) {
...
+			{Id: uuid.NewV4().String(), Spec: &storage.NetworkPolicySpec{PodSelector: &storage.LabelSelector{MatchLabels: map[string]string{"app": "match"}}}},
+			{Id: uuid.NewV4().String(), Spec: &storage.NetworkPolicySpec{PodSelector: &storage.LabelSelector{MatchLabels: map[string]string{"nomatch": "nomatch"}}}},
+		}, nil)
+	s := serviceImpl{
+		netpols: mockNetpolDS,
+	}
+	eCtx := enricher.EnrichmentContext{Namespace: "ns", ClusterID: uuid.NewV4().String()}
+	d := storage.Deployment{Id: uuid.NewV4().String(), PodLabels: map[string]string{"app": "match"}}
+
+	actual, err := s.getAppliedNetpolsForDeployment(context.Background(), eCtx, &d)
```
Comment (vikin91) [inline central/detection/service/service_impl_test.go:708]: [nit] I see again the var `actual` from that pattern you mentioned last time. I usually stick to `got` as it is named in unit tests generated automatically by VSCode. Do you know maybe whether go has an unified opinion on naming this var?

**Review by vikin91 — APPROVED** [review]
It actually looks really good! I am not the most fluent person in the Central detector, so I looked mainly at the rest of the code. I found some minor things that could be improved (in the comments), but no blockers.

File: central/detection/service/service_impl.go [hunk 7]
```diff
@@ -166,6 +171,7 @@ func (s *serviceImpl) DetectBuildTime(ctx context.Context, req *apiV1.BuildDetec
 		}
 
 		enrichmentContext.ClusterID = clusterID
+
```
Comment (fredrb) [inline central/detection/service/service_impl.go:174]: [nit] remove

File: central/detection/service/service_impl.go (hunk 3 above)
Comment (fredrb) [inline central/detection/service/service_impl.go:260]: Is `InvalidArgs` reasonable here? Could `netpols/GetNetworkPolicies` return an error that is unrelated to the parameters passed?

File: central/detection/service/service_impl.go [hunk 8]
```diff
@@ -352,6 +376,11 @@ func (s *serviceImpl) DetectDeployTimeFromYAML(ctx context.Context, req *apiV1.D
 		eCtx.ClusterID = clusterID
 	}
 
+	ns := req.GetNamespace()
+	if ns != "" {
+		eCtx.Namespace = ns
```
Comment (fredrb) [inline central/detection/service/service_impl.go:381]: Is `eCtx.Namespace` defaulted to `"default"`?

File: proto/api/v1/detection_service.proto [hunk 9]
```diff
@@ -43,6 +43,7 @@ message DeployYAMLDetectionRequest {
     repeated string policy_categories    = 4;
     // Cluster to delegate scan to, may be the cluster's name or ID.
     string          cluster              = 6;
+    string          namespace            = 7;
```
Comment (fredrb) [inline proto/api/v1/detection_service.proto:46]: This was added in #9110

File: central/detection/service/service_impl_test.go (hunk 6 above)
Comment (Maddosaurus) [reply central/detection/service/service_impl_test.go:708]: I've quickly checked the go source itself, as well as an[ article on naming conventions](https://ieftimov.com/posts/testing-in-go-naming-conventions/).
It seems that both patterns, `actual`/`expected` and `got`/`want` are used in the Go codebase itself as well as popular projects.
Maybe this is a good extension for our code guidelines. 
@rukletsov would you happen to have an informed opinion on this matter? 😄

File: central/detection/service/service_impl.go [hunk 10]
```diff
@@ -220,10 +225,19 @@ func (s *serviceImpl) enrichAndDetect(ctx context.Context, enrichmentContext enr
 		EnforcementOnly: enrichmentContext.EnforcementOnly,
 	}
 
+	var appliedNetpols *augmentedobjs.NetworkPoliciesApplied
+	appliedNetpols, err = s.getAppliedNetpolsForDeployment(ctx, enrichmentContext, deployment)
+	if err != nil {
+		log.Warnf("Could not find applied network policies for deployment %s. Continuing with deployment enrichment.", deployment.GetName())
```
Comment (vikin91) [inline central/detection/service/service_impl.go:231]: Sorry, looking at this after some break makes me see it from a different perspective :)

Would it make sense to include the `err` message in that log line? I don't know what kind of errors are returned from that func, but if there is more than one type of error possible, then I think it maybe useful to print it.

File: central/detection/service/service_impl_test.go (hunk 6 above)
Comment (vikin91) [reply central/detection/service/service_impl_test.go:709]: That is fine, I am resolving the comment, but feel free to wait on Alex'es opinion.

File: pkg/protoconv/resources/resources_test.go [hunk 11]
```diff
@@ -141,6 +141,21 @@ func TestCronJobPopulateSpec(t *testing.T) {
...
+		Spec: appsV1.DeploymentSpec{
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"app": "nginx"},
+				},
+				Spec: v1.PodSpec{Containers: []v1.Container{{Name: "nginx"}}},
+			},
+		},
+	}
+	d, _ := NewDeploymentFromStaticResource(deployment, "Deployment", "", "")
```
Comment (vikin91) [inline pkg/protoconv/resources/resources_test.go:155]: We are in test, so checking the error returned from a function is recommended and may save us from potential bugs in the future. Let's maybe add an assertion that the second param is nil.

**Review by vikin91 — APPROVED** [review]
Still good, but looking at it again after a while allowed me to find two more potential nits.

File: central/detection/service/service_impl.go (hunk 3 above)
Comment (fredrb) [reply central/detection/service/service_impl.go]: IMO this seems like a generic error, and we shouldn't try to fit into a generalized error box. Ideally, `GetNetworkPolicies` will return a error type based on the error that occured and we can just wrap with our message here. 

I suggest changing to:
```
errors.Wrapf(err, "failed to get network policies for cluster %s and namespace %s", cluster, namespace)
```

Comment (Maddosaurus) [reply central/detection/service/service_impl.go]: Fair, changed!

File: central/detection/service/service_impl.go (hunk 10 above)
Comment (Maddosaurus) [reply central/detection/service/service_impl.go:231]: Done!
